dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/WorkflowDefinitionServiceImpl.java

```diff
@@ -409,14 +409,16 @@ public void saveWorkflowLineage(long projectCode,
                                     long workflowDefinitionCode,
                                     int workflowDefinitionVersion,
                                     List<TaskDefinitionLog> taskDefinitionLogList) {
-        List<WorkflowTaskLineage> workflowTaskLineageList =
-                generateWorkflowLineageList(taskDefinitionLogList, workflowDefinitionCode, workflowDefinitionVersion);
-        if (workflowTaskLineageList.isEmpty()) {
-            return;
-        }
+        List<WorkflowTaskLineage> workflowTaskLineageList = generateWorkflowLineageList(taskDefinitionLogList,
+                workflowDefinitionCode, workflowDefinitionVersion);
 
-        int insertWorkflowLineageResult = workflowLineageService.updateWorkflowLineage(workflowTaskLineageList);
-        if (insertWorkflowLineageResult <= 0) {
+        int insertWorkflowLineageResult =
+                workflowLineageService.updateWorkflowLineage(workflowDefinitionCode, workflowTaskLineageList);
+        if (CollectionUtils.isEmpty(workflowTaskLineageList)) {
+            log.info(
+                    "Delete workflow lineage because current lineage is empty, projectCode: {}, workflowDefinitionCode: {}, workflowDefinitionVersion: {}",
+                    projectCode, workflowDefinitionCode, workflowDefinitionVersion);
+        } else if (insertWorkflowLineageResult <= 0) {
```

Comment:
```suggestion
       if (!CollectionUtils.isNotEmpty(workflowTaskLineageList) && insertWorkflowLineageResult <= 0) {
```
Remove the log here, this service should not be aware of delete operations.
In fact, throwing exceptions should be the responsibility of the DAO layer.

Reply:
As you suggested, I refactored: Instead of deleting the logs directly, I moved the lineage logic to WorkflowTaskLineageService.updateWorkflowLineage() and simplified WorkflowDefinitionService.saveWorkflowLineage().